pkg/allocator/allocator.go

```diff
@@ -636,6 +636,15 @@ func (a *Allocator) Allocate(ctx context.Context, key AllocatorKey) (idpool.ID,
 		select {
 		case <-ctx.Done():
 			scopedLog.WithError(ctx.Err()).Warning("Ongoing key allocation has been cancelled")
+
+			// If the deadline as exceeded and there was a backend error, this
+			// means that kvstore connectivity is unstable. We fatal Cilium in
+			// hopes of finding a new backend.
+			if _, backendErr := a.backend.Status(); backendErr != nil &&
+				ctx.Err() == context.DeadlineExceeded {
+				scopedLog.WithField("error", backendErr).Fatal("Key allocation failed with unstable kvstore connectivity")
```

Comment:
It will only fatal if an identity allocation actually times out. If the operator is only doing maintenance by upgrading one of the peers, the etcd cluster will remain healthy and it won't fatal.